batch/src/main/java/org/openmrs/analytics/JdbcFetchUtil.java

```diff
@@ -0,0 +1,205 @@
+		comboPooledDataSource.setUser(dbUser);
+		comboPooledDataSource.setPassword(dbPassword);
+		comboPooledDataSource.setMaxPoolSize(dbcMaxPoolSize);
+		comboPooledDataSource.setInitialPoolSize(10); // TODO: make this configurable if issues arises
+	}
+	
+	public static class FetchUuIds extends PTransform<PCollection<KV<String, Integer>>, PCollection<String>> {
+		
+		private final String tableName;
+		
+		private final JdbcIO.DataSourceConfiguration dataSourceConfig;
+		
+		public FetchUuIds(String tableName, JdbcIO.DataSourceConfiguration dataSourceConfig) {
+			this.tableName = tableName;
+			this.dataSourceConfig = dataSourceConfig;
+		}
+		
+		@Override
+		public PCollection<String> expand(PCollection<KV<String, Integer>> idRanges) {
+			String tableId = this.tableName + "_id";
+			return idRanges.apply(String.format("Read UUIDs from %s", this.tableName),
+			    JdbcIO.<KV<String, Integer>, String> readAll().withDataSourceConfiguration(this.dataSourceConfig)
+			            .withParameterSetter(new JdbcIO.PreparedStatementSetter<KV<String, Integer>>() {
+				            
+				            @Override
+				            public void setParameters(KV<String, Integer> element, PreparedStatement preparedStatement)
+				                    throws Exception {
+					            String[] range = Objects.requireNonNull(element.getKey()).split(",");
+					            preparedStatement.setInt(1, Integer.parseInt(range[0]));
+					            preparedStatement.setInt(2, Integer.parseInt(range[1]));
+				            }
+			            }).withOutputParallelization(false)
+			            .withQuery(
+			                String.format("select uuid from %s where %s >= ? and %s < ?", this.tableName, tableId, tableId))
+			            .withRowMapper(new JdbcIO.RowMapper<String>() {
+				            
+				            @Override
+				            public String mapRow(ResultSet resultSet) throws Exception {
+					            return resultSet.getString("uuid");
+				            }
+			            }).withCoder(NullableCoder.of(StringUtf8Coder.of())));
+		}
+	}
+	
+	/**
+	 * batch together request using batchSize and generate segment descriptors `_id?<uuid>,<uuid>..`
+	 */
+	public static class CreateSearchSegments extends PTransform<PCollection<String>, PCollection<SearchSegmentDescriptor>> {
+		
+		private final String baseBundleUrl;
+		
+		private final String resourceType;
+		
+		private final int batchSize;
+		
+		public CreateSearchSegments(String resourceType, String baseBundleUrl, int batchSize) {
+			this.baseBundleUrl = baseBundleUrl;
+			this.batchSize = batchSize;
+			this.resourceType = resourceType;
+		}
+		
+		@Override
+		public PCollection<SearchSegmentDescriptor> expand(PCollection<String> uuIds) {
+			return uuIds
+			        // create KV required by GroupIntoBatches
+			        .apply(ParDo.of(new DoFn<String, KV<String, String>>() {
+				        
+				        @ProcessElement
+				        public void processElement(ProcessContext c) {
```

Comment:
Is there any reason to prefer `ProcessContext` to `@Element String ...` here (like all other `DoFn`s)?

Reply:
I have switched back to `@Element String ...`  to make it more consistent. Thanks!